ethereum/trin
# add upnp support

### What was wrong?

This PR adds UPnP support.  UPnP is the default NAT used in libp2p.  UPnP protocol will ask the gateway to map a private IP/port to an external IP/port and return the external address to Trin.  By encoding the external address in ENR, Trin will allow any node to access a Trin node running behind the gateway.

Test:
- Run a Trin node behind a gateway with an empty bootnode.  The node will print the ENR with an external IP/port mapped by UPnP.
- Run another Trin node using the bootnode's ENR, which connects the bootnode.
- Call portal_historyGossip (using test vector) from one node and check that another node receives the data.

### How was it fixed?

### To-Do

[//]: # (Stay ahead of things, add list items here!)
- [ ] Clean up commit history and use [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/).
